humility-cmd/src/hiffy.rs

```diff
@@ -1006,3 +1011,147 @@ impl<'a> HiffyContext<'a> {
+}
+
+/// This is a reimplementation of `fn send` in `task/hiffy/src/common.rs`.
+///
+/// We extract a handful of parameters, then pack them into a packet and send
+/// them across the network to the `udprcp` task.
+fn hiffy_send_fn(
+    stack: &[Option<u32>],
+    _data: &[u8],
+    rval: &mut [u8],
+) -> Result<usize, Failure> {
+    let mut payload = [0u8; 32];
+
+    let sp = stack.len();
+    if sp < 4 {
+        return Err(Failure::Fault(Fault::MissingParameters));
+    }
+
+    let nreply =
+        stack[sp - 1].ok_or(Failure::Fault(Fault::EmptyParameter(4)))?;
+    let nbytes: u16 = stack[sp - 2]
+        .ok_or(Failure::Fault(Fault::EmptyParameter(3)))?
+        .try_into()
+        .map_err(|_| Failure::Fault(Fault::BadParameter(1)))?;
+
+    let fp = sp - (nbytes as usize + 4);
+
+    let task: u16 = stack[fp]
+        .ok_or(Failure::Fault(Fault::EmptyParameter(0)))?
+        .try_into()
+        .map_err(|_| Failure::Fault(Fault::BadParameter(1)))?;
+
+    let op: u16 = stack[fp + 1]
+        .ok_or(Failure::Fault(Fault::EmptyParameter(1)))?
+        .try_into()
+        .map_err(|_| Failure::Fault(Fault::BadParameter(1)))?;
+
+    let base = fp + 2;
+
+    for i in base..base + nbytes as usize {
+        payload[i - base] = stack[i]
+            .ok_or(Failure::Fault(Fault::EmptyParameter(2)))?
+            .try_into()
+            .map_err(|_| Failure::Fault(Fault::BadParameter(2)))?;
+    }
+
+    let mut buf = [0u8; 1024]; // matches buffer size in `task-udprpc`
+    HIFFY_SEND_WORKSPACE.with(|workspace| {
+        let mut workspace = workspace.borrow_mut();
+        let (hubris, core) = {
+            // SAFETY: we only ever call this function when the pointers are
+            // populated, and reset them to null / None afterwards.  This means
+            // we should fail at this unwrap() if someone violates the rules.
+            unsafe {
+                (
+                    workspace.hubris.unwrap().as_ref(),
+                    workspace.core.unwrap().as_mut(),
+                )
+            }
+        };
+        let image_id = hubris.image_id().unwrap();
+
+        let header = RpcHeader {
+            image_id: U64::from_bytes(image_id.try_into().unwrap()),
+            task: U16::new(task),
+            op: U16::new(op),
+            nreply: U16::new(nreply as u16),
+            nbytes: U16::new(nbytes),
+        };
+
+        let mut packet = header.as_bytes().to_vec();
+        packet.extend(&payload[0..nbytes as usize]);
+
+        // Send the packet out
+        if let Err(e) = core.send(&packet) {
+            workspace.errors.push(e);
+            return Err(Failure::FunctionError(0));
+        }
+
+        // Try to receive a reply
+        match core.recv(buf.as_mut_slice()) {
+            Err(e) => {
+                workspace.errors.push(e);
+                return Err(Failure::FunctionError(0));
+            }
+            Ok(_n) => (),
+        }
+
+        workspace.results.push(buf.to_vec());
```

Comment:
Should this be `buf[..n].to_vec()` (at which point maybe it moves into the `match` above?

Reply:
This matches the previous code, probably because the decoder allows for trailing data in the slice.  Still, it would probably be cleaner to just push the valid packet data, so I'll do that.